unittests/AST/StructuralEquivalenceTest.cpp

```diff
@@ -209,5 +209,165 @@ TEST_F(StructuralEquivalenceTest, WrongOrderOfFieldsInClass) {
+                          "struct foo { void x(char); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, ParamConst) {
+  auto t = makeNamedDecls("struct foo { void x(int&); };",
+                          "struct foo { void x(const int&); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, ParamName) {
+  auto t = makeNamedDecls("struct foo { void x(int a); };",
+                          "struct foo { void x(int b); };", Lang_CXX);
+  EXPECT_TRUE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Virtual) {
+  auto t = makeNamedDecls("struct foo { void x(); };",
+                          "struct foo { virtual void x(); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Pure) {
+  auto t = makeNamedDecls("struct foo { virtual void x(); };",
+                          "struct foo { virtual void x() = 0; };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Const) {
+  auto t = makeNamedDecls("struct foo { void x(); };",
+                          "struct foo { void x() const; };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Static) {
+  auto t = makeNamedDecls("struct foo { void x(); };",
+                          "struct foo { static void x(); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Variadic) {
+  auto t = makeNamedDecls("struct foo { void x(int x...); };",
+                          "struct foo { void x(int x); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Ptr) {
+  auto t = makeNamedDecls("struct foo { void x(int *); };",
+                          "struct foo { void x(int); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Default) {
+  auto t = makeNamedDecls("struct foo { foo(); };",
+                          "struct foo { foo() = default; };", Lang_CXX11);
+  EXPECT_TRUE(testStructuralMatch(get<0>(t), get<1>(t)));
+  // "defaulted" affects the implementation
+  // not checked by "structural equivalence"
+  // (and checking it causes unexpected other problems)
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Delete) {
+  auto t = makeNamedDecls("struct foo { void x(); };",
+                          "struct foo { void x() = delete; };", Lang_CXX11);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Constructor) {
+  auto t = makeNamedDecls("struct foo { };",
+                          "struct foo { foo(); };", Lang_CXX);
```

Comment:
In this case it is not exactly clear which Decl is matched by the lookup inside `makeNamedDecls`. Is it the CXXRecordDecl or the CXXConstructorDecl? I think we could reuse `ASTImporterTestBase::getTuDecl()` here so we could explicitly match for the Decls we want to compare. (For that, we need a common TestBase class).

Reply:
I this case it should be the struct foo because a lookup is performed on the top level DeclContext. There is an assertion that asserts if multiple results are found.

Reply:
OK.